R/teal_data_utils.R

```diff
@@ -42,8 +42,14 @@ NULL
 .subset_teal_data <- function(data, datanames) {
   checkmate::assert_class(data, "teal_data")
   checkmate::assert_class(datanames, "character")
-  datanames_corrected <- intersect(datanames, ls(data@env))
-  dataname_corrected_with_raw <- intersect(c(datanames, sprintf("%s._raw_", datanames)), ls(data@env))
+  datanames_corrected <- intersect(datanames, .teal_data_ls(data))
+  dataname_corrected_with_raw <- c(
+    datanames_corrected,
+    intersect(
+      sprintf(".%s_raw", datanames_corrected),
+      ls(data@env, all.names = TRUE)
+    )
+  )
```

Comment:
I am trying to figure out why in `.subset_teal_data` we once use `.teal_data_ls(data)` and in the next line we use `ls(data@env, all.names = TRUE)`.

Currently the only difference between `.teal_data_ls` and `ls`, is that the first uses `include_parent_datanames` which extracts the names of the `parents` for datasets based on `join_keys`.

Is the point of `include_parent_datanames` in `.teal_data_ls` to extend the list of names by `parents` or just to reorder the names? Based on the below code, where `iris` is a parent of `iris_join` I don't get `iris` in `@env` as a result of `teal:::.subset_teal_data(data, "iris_join"` but I do see that `join_keys` are created for that case.

I am asking because I am a bit confused. I see the same behavior on main but without the need to have `include_parent_datanames` in `.teal_data_ls`

# This branch

```r
iris_join <- data.frame(Species = unique(iris$Species), Letter = letters[1:3])

data <- teal.data::teal_data(
  iris = iris, .iris_raw = iris, 
  iris_join = iris_join, .iris_join_raw = iris_join, 
  mtcars = mtcars, .mtcars_raw = mtcars, 
  x = swiss,
  factor = factor(c('a', 'b'))
)
teal.data::datanames(data) <- c('iris', 'iris_join', 'mtcars')


teal.data::join_keys(data) <- teal.data::join_keys(
  teal.data::join_key("iris", "iris_join", keys = c("Species"))
)


teal:::.subset_teal_data(data, 'iris')
#> Registered S3 method overwritten by 'teal':
#>   method        from      
#>   c.teal_slices teal.slice
#> ✖ unverified teal_data object
#> <environment: 0x0000029ab3151490> [L]
#> Parent: <environment: package:stats>
#> Bindings:
#> • iris: <df[,5]> [L]
#> • .iris_raw: <df[,5]> [L]
teal:::.subset_teal_data(data, 'iris_join')
#> ✖ unverified teal_data object
#> <environment: 0x0000029ab5a54680> [L]
#> Parent: <environment: package:stats>
#> Bindings:
#> • iris_join: <df[,2]> [L]
#> • .iris_join_raw: <df[,2]> [L]
teal.data::join_keys(teal:::.subset_teal_data(data, 'iris'))
#> An empty join_keys object.
teal.data::join_keys(teal:::.subset_teal_data(data, 'iris_join'))
#> A join_keys object containing foreign keys between 2 datasets:
#> iris: [no primary keys]
#>   <-- iris_join: [Species]
#> iris_join: [no primary keys]
#>   --> iris: [Species]


teal:::include_parent_datanames('iris', teal.data::join_keys(data))
#> [1] "iris"
teal:::include_parent_datanames('iris_join', teal.data::join_keys(data))
#> [1] "iris"      "iris_join"
```


# Main branch

```r
iris_join <- data.frame(Species = unique(iris$Species), Letter = letters[1:3])

data <- teal.data::teal_data(
  iris = iris, iris._raw_ = iris,
  iris_join = iris_join, iris_join._raw_ = iris_join,
  mtcars = mtcars, mtcars._raw_ = mtcars,
  x = swiss,
  factor = factor(c('a', 'b'))
)

teal.data::datanames(data) <- c('iris', 'iris_join', 'mtcars')


teal.data::join_keys(data) <- teal.data::join_keys(
  teal.data::join_key("iris", "iris_join", keys = c("Species"))
)


teal:::.subset_teal_data(data, 'iris')
#> Registered S3 method overwritten by 'teal':
#>   method        from      
#>   c.teal_slices teal.slice
#> ✖ unverified teal_data object
#> <environment: 0x0000022e5af49b20> [L]
#> Parent: <environment: package:stats>
#> Bindings:
#> • iris._raw_: <df[,5]> [L]
#> • iris: <df[,5]> [L]
teal:::.subset_teal_data(data, 'iris_join')
#> ✖ unverified teal_data object
#> <environment: 0x0000022e5d7c12c8> [L]
#> Parent: <environment: package:stats>
#> Bindings:
#> • iris_join: <df[,2]> [L]
#> • iris_join._raw_: <df[,2]> [L]
teal.data::join_keys(teal:::.subset_teal_data(data, 'iris'))
#> An empty join_keys object.
teal.data::join_keys(teal:::.subset_teal_data(data, 'iris_join'))
#> A join_keys object containing foreign keys between 2 datasets:
#> iris: [no primary keys]
#>   <-- iris_join: [Species]
#> iris_join: [no primary keys]
#>   --> iris: [Species]
```

Reply:
Ok, thanks, all clear!